kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/ProxyChannelStateMachineTest.java

```diff
@@ -0,0 +1,698 @@
+        proxyChannelStateMachine.onServerUnwritable();
+
+        // When
+        proxyChannelStateMachine.onServerUnwritable();
+
+        // Then
+        verify(frontendHandler, times(1)).applyBackpressure();
+    }
+
+    @Test
+    void shouldUnblockClientReads() {
+        // Given
+        stateMachineInClientActive();
+        proxyChannelStateMachine.clientReadsBlocked = true;
+        proxyChannelStateMachine.onServerWritable();
+
+        // When
+        proxyChannelStateMachine.onServerWritable();
+
+        // Then
+        verify(frontendHandler, times(1)).relieveBackpressure();
+    }
+
+    @Test
+    void shouldBlockServerReads() {
+        // Given
+        stateMachineInForwarding();
+        proxyChannelStateMachine.onClientUnwritable();
+
+        // When
+        proxyChannelStateMachine.onClientUnwritable();
+
+        // Then
+        verify(backendHandler, times(1)).applyBackpressure();
+    }
+
+    @Test
+    void shouldCloseOnClientRuntimeException() {
+        // Given
+        stateMachineInForwarding();
+        RuntimeException cause = new RuntimeException("Oops!");
+
+        // When
+        proxyChannelStateMachine.onClientException(cause, true);
+
+        // Then
+        assertThat(proxyChannelStateMachine.state()).isInstanceOf(ProxyChannelState.Closed.class);
+        verify(backendHandler).inClosed();
+        verify(frontendHandler).inClosed(ArgumentMatchers.notNull(UnknownServerException.class));
+    }
+
+    @Test
+    void shouldCloseOnClientFrameOversizedException() {
+        // Given
+        stateMachineInForwarding();
+        RuntimeException cause = new DecoderException(new FrameOversizedException(2, 1));
+
+        // When
+        proxyChannelStateMachine.onClientException(cause, true);
+
+        // Then
+        assertThat(proxyChannelStateMachine.state()).isInstanceOf(ProxyChannelState.Closed.class);
+        verify(backendHandler).inClosed();
+        verify(frontendHandler).inClosed(ArgumentMatchers.notNull(InvalidRequestException.class));
+    }
+
+    @Test
+    void shouldCloseOnServerRuntimeException() {
+        // Given
+        stateMachineInForwarding();
+        RuntimeException cause = new RuntimeException("Oops!");
+
+        // When
+        proxyChannelStateMachine.onServerException(cause);
+
+        // Then
+        assertThat(proxyChannelStateMachine.state()).isInstanceOf(ProxyChannelState.Closed.class);
+        verify(backendHandler).inClosed();
+        verify(frontendHandler).inClosed(cause);
+    }
+
+    @Test
+    void shouldUnblockServerReads() {
+        // Given
+        stateMachineInForwarding();
+        proxyChannelStateMachine.serverReadsBlocked = true;
+
+        // When
+        proxyChannelStateMachine.onClientWritable();
+        proxyChannelStateMachine.onClientWritable();
+
+        // Then
+        verify(backendHandler, times(1)).relieveBackpressure();
+    }
+
+    @Test
+    void inStartStateShouldCallInClientActive() {
```

Comment:
feel like the naming is missing something, it's about the result of a transition from start to active. Maybe a pattern like `handlerNotifiedOnTransitionFromStartToClientActive`?

Reply:
resolved in 27697763bed9df88046ee800ef2d95ab71471921